Add schema for new artifact upload configuration

odyssey4me opened the pull request:
In order to facilitate in-repo configuration of artifacts
to be uploaded from jobs, we add the schema validation and
appropriate tests for the new 'artifacts' key and the two
initial types: 'log' and 'file'.

Issue: [RE-1325](https://rpc-openstack.atlassian.net/browse/RE-1325)

Commit:
Add schema for new artifact upload configuration

In order to facilitate in-repo configuration of artifacts
to be uploaded from jobs, we add the schema validation and
appropriate tests for the new 'artifacts' key and the two
initial types: 'log' and 'file'.

JIRA: RE-1325

dealako commented:
I'll defer to others to review - I'm not familiar with the desired schema or the python library used to create/validate the schema.

Commit:
Add new command to get/validate component metadata

In order to be able to easily verify whether component
metadata has been correctly configured, we add a new
command to 'get' the metadata, which will also verify
it before outputting it.

JIRA: RE-1325

hughsaunders reviewed (commented):
Looks ok, but file and log are quite similar, after all logs are usually files, so a note about the use cases for each would be useful. 

Whats the difference between a log and an file with expire_after: 30 days?

Also whats the unit for expire_after?

odyssey4me commented:
@hughsaunders The ``log`` type only has one parameter. The destination is hard-coded, as is the expiry. The ``file`` type will go to a different, consistent container. Perhaps this is better reviewed in tandem with the implementation: https://github.com/rcbops/rpc-gating/pull/979

hughsaunders reviewed (dismissed):
I see enough differentiation between file and log. Code looks ok, I'll leave the schema definition checks to @git-harry

git-harry commented on rpc_component/schemata.py:
```diff
@@ -134,20 +135,59 @@ def is_version_ids_unique(releases):
+
+artifacts_schema = Schema(
+    {
+        Optional("artifacts"): And(
```
I don't think you need the `And` because there is only the list.

odyssey4me replied:
👍

git-harry commented on rpc_component/schemata.py:
```diff
@@ -134,20 +135,59 @@ def is_version_ids_unique(releases):
+    {
+        Optional("artifacts"): And(
+            [
+                Or(artifacts_file_schema, artifacts_log_schema, None)
```
What does `None` represent? How does that compare with an empty list?

odyssey4me replied:
``None`` is to allow for the list to be empty, this this is to cater for:

```yaml
artifacts: []
```

odyssey4me replied:
Looks like it works without it, so I've removed it.

git-harry commented on rpc_component/schemata.py:
```diff
@@ -134,20 +135,59 @@ def is_version_ids_unique(releases):
+        "type": "file",
+        "source": And(str, len),
+        Optional("dest"): And(str, len),
+        Optional("expire_after"): And(str, len)
```
I assume this must be in a specific format, you can make that part of the schema so that it is validated. For example, if it's a number of days you could use a function to check it's an integer greater than 0 or you could use a regex or maybe something from the `datetime` class.

odyssey4me replied:
Good call, done.

git-harry commented on rpc_component/cli.py:
```diff
@@ -242,6 +242,42 @@ def dependency(components_dir, **kwargs):
+    new_metadata = {"artifacts": [], "dependencies": []}
+    try:
+        metadata = c_lib.load_data(filepath)
+    except yaml.parser.ParserError as e:
```
I think it would be better to put this in the `c_lib.load_data` function.

odyssey4me replied:
Fair enough.

git-harry commented on rpc_component/cli.py:
```diff
@@ -422,6 +458,18 @@ def parse_args(args):
+
+    metadata_get_parser = metadata_subparsers.add_parser(
+        "get",
+        help=("Validate the output the component metadata."),
```
You don't need these brackets.

odyssey4me replied:
👍

git-harry commented on rpc_component/cli.py:
```diff
@@ -242,6 +242,42 @@ def dependency(components_dir, **kwargs):
+    if subparser == "get":
+        try:
+            return s_lib.component_metadata_schema.validate(metadata)
+        except SchemaError as e:
```
I'm not sure this makes it more readable, the schema errors already tend to tell you as much, but if you want to try and improve how these errors present I think it would be better to do it where they are caught at the end of the file (line 506) so everything benefits.

odyssey4me replied:
Allowing the global exception to do this instead totally makes sense now that you suggest it.

git-harry commented on rpc_component/schemata.py:
```diff
@@ -134,20 +135,58 @@ def is_version_ids_unique(releases):
+        "type": "file",
+        "source": And(str, len),
+        Optional("dest"): And(str, len),
+        Optional("expire_after"): And(Use(int), lambda n: n > 0)
```
I would rather it could only be one type, this potentially causes a pain point due to the way each is handled as well as potentially causing the input to differ from the output depending on internal representation.

odyssey4me replied:
Fair enough, done.

git-harry reviewed (dismissed):
I would like to see the comment addressed but it will work as is so I've approved.
